### PR TITLE
[Group 4] Enable nullable annotations for `Microsoft.Extensions.Configuration.Binder`

### PR DESCRIPTION
**Pull request by maxkoshevoi**
Related to #43605, #54012

Annotated according to:
- https://github.com/dotnet/runtime/pull/57401

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs
Comment (maxkoshevoi) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:88]: Maybe it's better to remove null-check here? It would be easier to call `Get` with `null` than with `_ => { }`

**Comment by eerhardt** [comment]
@maxkoshevoi - let's concentrate on getting the Group 1, 2 and 3 libraries completed. Since it takes a lot of effort to review these PRs, and they have been sitting for a while, what do you think about closing the "Group 4" PRs for now?

**Comment by maxkoshevoi** [comment]
@eerhardt What's the harm in keeping them open? They won't be reviewed until parent PRs are merged anyway (#57401 in this case). Or do you just want to do it for some internal metric?
I can convert Group 4 PRs into drafts. That also should do the trick.

**Comment by eerhardt** [comment]
> What's the harm in keeping them open? Or do you just want to do it for some internal metric?

We do have an effort of trying to keep on top of our open PRs. As you can see, we currently have over 220 open PRs, some of them have been open for months with no activity.

The harm is someone coming to the PR and trying to review it, but it isn't ready to be reviewed. It also affects managing the PRs ("these ones can be reviewed" vs. "these ones are waiting for something").

**Comment by eerhardt** [comment]
> I can convert Group 4 PRs into drafts.

I think that would be great for now. Thanks!

File: src/libraries/Microsoft.Extensions.Configuration.Binder/ref/Microsoft.Extensions.Configuration.Binder.csproj
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/ref/Microsoft.Extensions.Configuration.Binder.csproj:3]:
```suggestion
    <TargetFrameworks>$(NetCoreAppCurrent);$(NetCoreAppMinimum);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
```

This applies to both .csproj files. Since 6.0 is LTS, we should be building a netcoreapp version for 6.0, as well as 7.0.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/ref/Microsoft.Extensions.Configuration.Binder.csproj
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/ref/Microsoft.Extensions.Configuration.Binder.csproj:20]: Is this necessary? Shouldn't we get this "for free" from the ProjectReference to `Microsoft.Extensions.Configuration.Abstractions.csproj`?

I tried removing this locally, and it seemed to build fine.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:33]: We shouldn't be changing implementation during nullable changes. Can you undo this?

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:70]: We shouldn't be changing implementation during nullable changes. Can you undo this?

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:210]: Is this refactoring necessary? It makes reviewing the nullable annotations much harder.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:510]: If there is no `Add` method, this will now silently no longer do anything. Is that what we want to happen?

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:651]: Is this change necessary? It seems to be changing behavior.

Maybe the `Type type` parameter shouldn't be nullable?

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:510]: We didn't seem to do anything before (there's an empty catch on the line bellow).

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks for the contribution here, @maxkoshevoi.

I pushed a change that should allow CI to pass successfully. Will merge when the CI is green.

**Pull request merged**
